### PR TITLE
Added caching for PhpParser result

### PR DESCRIPTION
**Pull request by POPSuL**
Please, see full description there https://github.com/ApiGen/ApiGen/issues/951#issuecomment-315245749

### CODE REVIEW

File: src/SourceLocator/Ast/Locator.php [hunk 1]
```diff
@@ -27,6 +27,11 @@ class Locator
      */
     private $parser;
 
+    /**
+     * @var array|string[]
+     */
+    private static $cache = [];
```
Comment (Ocramius) [inline src/SourceLocator/Ast/Locator.php]: Move this to a `CachedLocator`. Adding a cache, and ESPECIALLY a static one is a no-go here.

File: src/SourceLocator/Ast/CachedLocator.php [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+                $this->cache[$id] = $this->parser->parse($locatedSource->getSource());
+            }
+
+            return $this->findReflectionsInTree->__invoke(
+                $reflector,
+                $this->cache[$id],
+                $identifierType,
+                $locatedSource
+            );
+        } catch (\Exception $exception) {
```
Comment (Ocramius) [inline src/SourceLocator/Ast/CachedLocator.php]: Not needed

File: src/SourceLocator/Ast/CachedLocator.php [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+namespace Roave\BetterReflection\SourceLocator\Ast;
+
+use Roave\BetterReflection\Identifier\IdentifierType;
+use Roave\BetterReflection\Reflector\Reflector;
+use Roave\BetterReflection\SourceLocator\Located\LocatedSource;
+
+/**
+ * @internal
+ */
+class CachedLocator extends Locator
```
Comment (Ocramius) [inline src/SourceLocator/Ast/CachedLocator.php]: Let's extract an interface instead of using inheritance

File: src/SourceLocator/Ast/Locator.php [hunk 4]
```diff
@@ -20,12 +20,12 @@ class Locator
     /**
      * @var FindReflectionsInTree
      */
-    private $findReflectionsInTree;
+    protected $findReflectionsInTree;
```
Comment (Ocramius) [inline src/SourceLocator/Ast/Locator.php]: This should be reverted after we extract the interface

File: src/SourceLocator/Ast/Locator.php [hunk 5]
```diff
@@ -20,12 +20,12 @@ class Locator
...
 
     /**
      * @var Parser
      */
-    private $parser;
+    protected $parser;
```
Comment (Ocramius) [inline src/SourceLocator/Ast/Locator.php]: This should be reverted after we extract the interface

File: src/SourceLocator/Type/ComposerSourceLocator.php [hunk 6]
```diff
@@ -21,9 +22,9 @@ class ComposerSourceLocator extends AbstractSourceLocator
      */
     private $classLoader;
 
-    public function __construct(ClassLoader $classLoader)
+    public function __construct(ClassLoader $classLoader, AstLocator $locator = null)
```
Comment (Ocramius) [inline src/SourceLocator/Type/ComposerSourceLocator.php]: We can make the argument required. What is the impact of that?

Comment (Ocramius) [inline src/SourceLocator/Type/ComposerSourceLocator.php]: As on: how much useeland code would need changing?

File: src/SourceLocator/Ast/Locator.php (hunk 5 above)
Comment (burzum) [inline src/SourceLocator/Ast/Locator.php]: Why does it have to be private?

File: src/SourceLocator/Ast/Locator.php (hunk 4 above)
Comment (burzum) [inline src/SourceLocator/Ast/Locator.php]: Why does it have to be private?

**Comment by burzum** [comment]
I hope this gets finished and merged any time soon, because it's a great improvement!

**Comment by Ocramius** [comment]
@burzum I think it will not land in `2.0.0`. For now, we focused on features supported, while we may focus on performance for `2.1.0` or `3.0.0` if BC breaks are required for that to happen.

**Comment by TomasVotruba** [comment]
@POPSuL Will you have time to process the notes? Thank you.

**Comment by TomasVotruba** [comment]
I guess this will be replaced by: https://github.com/Roave/BetterReflection/pull/312

**Comment by Ocramius** [comment]
@TomasVotruba I'm still not sure about the approach. I will likely end building a `BetterReflectionKernel` that implements a PSR-11 interface or equivalent, and then is passed around through components.

I definitely need to get rid of the staticness in the lib itself.

**Comment by TomasVotruba** [comment]
@ocramius That would be great. I use this approach in my app for years and I'm surprised no other package is using DI (is that antipattern? :))

Was that some intention on this packages?

I had to register and setup all services manually using Symfony DI. If that would be by default here, that would be awesome.

**Comment by Ocramius** [comment]
Replaced (after much much re-design work) in https://github.com/Roave/BetterReflection/pull/338

**Comment by TomasVotruba** [comment]
Thanks for the update

**Pull request closed without merging**